_posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md

```diff
@@ -0,0 +1,248 @@
+
+K8s deployments are highly elastic and can be automatically scaled up and down via Horizontal Pod Autoscale depending on the amount of traces flowing through the pipeline.
+
+![tracing_step2](/assets/media/blog-images/2021-12-02-distributed-tracing-pipeline-with-opentelemetry/tracing_step2.png){: .img-fluid}
+
+
+The agent-gateway architecture also allows us to use [Tail Sampling Processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/tailsamplingprocessor/README.md) with more availability in this setup. Tail Sampling Processor enables us to make more intelligent choices when it comes to sampling traces. This is especially true for latency measurements, which can only be measured after they are complete. Since the collector sits at the end of the pipeline and has a complete picture of the distributed trace, sampling determinations are made in opentelemetry collector to sample based on isolated, independent portions of the trace data.
+
+Today, this processor only works with a single instance of the collector. The workaround is to utilize [Trace ID aware load balancing](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/loadbalancingexporter/README.md) to support multiple collector instances and avoid a single point of failure. This load balancer exporter is added to the agent configuration. It is responsible for consistently exporting spans and logs belonging to a trace to the same backend gateway collector for tail based sampling.
+
+### Filtering traces for tail based sampling
+
+We will use a combination of filters to sample the traces. The filters for tail sampling are positive selections only, so if a trace is caught by any of the filter, it will be sampled; Alternatively, if no filter catches a trace, then it will not be sampled. The filters for tail based sampling can be chained together to get the desired effect:
+
+- **latency**: Sample based on the duration of the trace. The duration is determined by looking at the earliest start time and latest end time, without taking into consideration what happened in between.
+- **probabilistic**: Sample a percentage of traces.
+- **status_code**: Sample based upon the status code (`OK`, `ERROR` or `UNSET`)
+- **rate_limiting**: Sample based on rate of spans per trace per second
+- **string_attribute**: Sample based on string attributes value matches, both exact and regex value matches are supported
+
+Since there is no blocklist for filters, we will be using a workaround to sample all but cherrypick services that we do not want sampled. we will be using the `string_attribute` filter to create a negative filter. The filter will be dropping specific service traces that we do not want sampled and sample all other traces. This can be done using the following:
+
+- **&#8594;** In the agent config, add a span attribute for all traces with a key/value pair (e.g - retain_span/false) using attribute processor.
```

Comment:
drop the leading `-` on the items in this list - it causes a double bullet:

<img width="896" alt="Screen Shot 2021-12-07 at 10 00 39 AM" src="https://user-images.githubusercontent.com/1152927/145073167-4dbb6354-f751-43ea-891e-7adbf2a61862.png">

In the architecture section, the arrows work nicely, so follow that formatting pattern
<img width="900" alt="Screen Shot 2021-12-07 at 10 01 21 AM" src="https://user-images.githubusercontent.com/1152927/145073262-b4764262-8a08-4cbd-bdce-ef849fd4f3d8.png">

Reply:
✔️